### PR TITLE
Rewards distribution

### PR DESCRIPTION
**Pull request by 0xAshish**
- rewards distribution accounting is done on Heimdall  layer
- with each checkpoint updated Merkle root of accounts is submitted
- in order to claim rewards, validators submits proof of accounting

### CODE REVIEW

File: contracts/staking/StakeManager.sol [hunk 1]
```diff
@@ -165,6 +175,31 @@ contract StakeManager is Validator, IStakeManager, RootChainable, Lockable {
...
+    address _contract = validators[validatorId].contractAddress;
+    if (_contract == address(0x0)) {
+      validators[validatorId].reward = validators[validatorId].reward.add(_reward);
+    }
+    else {
+      // TODO: delegator bond/share rate if return needs to be updated periodically
+      // otherwise validator can delay and get all the delegators reward
+      ValidatorContract(_contract).updateRewards(_reward, currentEpoch, validators[validatorId].amount);
+    }
+    totalRewardsLiquidated += validators[validatorId].reward;
```
[the remark this answers is not in the document]
Comment (0xAshish) [reply contracts/staking/StakeManager.sol]: yeah yeah, I did make this change before reading this comment.
Nice catch  @mankenavenkatesh Thanks

File: contracts/staking/StakeManager.sol [hunk 2]
```diff
@@ -165,6 +175,31 @@ contract StakeManager is Validator, IStakeManager, RootChainable, Lockable {
     emit ReStaked(validatorId, validators[validatorId].amount, totalStaked);
   }
 
+  function claimRewards(uint256 validatorId, uint256 accountBalance, uint256 index, bytes memory proof) public /*onlyStaker(validatorId) */ {
+    // accountState = keccak256(abi.encodePacked(validatorId, accountBalance))
+    require(keccak256(abi.encodePacked(validatorId, accountBalance)).checkMembership(index, accountStateRoot, proof));
```
[the remark this answers is not in the document]
Comment (0xAshish) [reply contracts/staking/StakeManager.sol]: I had this in mind but the assumption is that it's transactors money.

**Comment by atvanguard** [comment]
Has `challengeStateRootUpdate` been discussed? Going back to _challengeable_ checkpoints will change the system assumptions significantly. @mankenavenkatesh @jdkanani

**Comment by 0xAshish** [comment]
> Has `challengeStateRootUpdate` been discussed? Going back to _challengeable_ checkpoints might change the system assumptions significantly. @mankenavenkatesh @jdkanani

Yup, so the assumption is that after `challengeStateRootUpdate` will stop the checkpoints and only once the account state is updated with more threshold like 80-90% or even more 100% votes then only will restart the checkpoints.
To be clear this worst of worst-case where 2/3+1 colluded, we need to slash all 2/3+1 once the `challengeStateRootUpdate` has been succeeded.
also, the challenge period is a single checkpoint. since this is validators own rewards at stake they should verify it at Heimdall and mainnet layer.
Still figuring out the inner details any inputs would be appreciated 😄

**Comment by atvanguard** [comment]
> The challenge period is a single checkpoint.
This is the culprit. I think this period is far too low.
Attack scenario: Other validators notice a validator is down, they make his account balance 0, then expectation is validator has to come back online within a single checkpoint.

If you think about about having longer challenge durations, what happens to checkpoints made after the malicious one needs to be thought about and how will plasma exits be affected (if at all).

**Comment by vaibhavchellani** [comment]
>Attack scenario: Other validators notice a validator is down, they make his account balance 0, then expectation is validator has to come back online within a single checkpoint.

The assumption here is that 2/3+1 validators collude. It should be quite possible that people can challenge anytime. Also this model relies on 2/3+1 security so this attack is possible for sure and there will be fund loss. Having a challenging period of 1 checkpoint is something we don't `have` to do but is just a fallback for very small validators.

**Comment by 0xAshish** [comment]
@atvanguard please push this PR by today, thanks 🌚

File: contracts/staking/StakeManager.sol [hunk 3]
```diff
@@ -165,6 +176,40 @@ contract StakeManager is Validator, IStakeManager, RootChainable, Lockable {
...
+    if (_contract == address(0x0)) {
+      validators[validatorId].reward = validators[validatorId].reward.add(_reward);
+    }
+    else {
+      // TODO: delegator bond/share rate if return needs to be updated periodically
+      // otherwise validator can delay and get all the delegators reward
+      ValidatorContract(_contract).updateRewards(_reward, currentEpoch, validators[validatorId].amount);
+    }
+    totalRewardsLiquidated += _reward;
+    require(totalRewardsLiquidated <= totalRewards, "Oops this shouldn't have happened");// pos 2/3+1 is colluded
```
Comment (atvanguard) [inline contracts/staking/StakeManager.sol]: Error message is inconclusive.

File: contracts/staking/StakeManager.sol [hunk 4]
```diff
@@ -165,6 +176,40 @@ contract StakeManager is Validator, IStakeManager, RootChainable, Lockable {
...
+    uint256 amount = validators[validatorId].reward;
+    address _contract = validators[validatorId].contractAddress;
+    if (_contract != address(0x0)) {
+      amount = amount.add(ValidatorContract(_contract).withdrawRewardsValidator());
+    }
+    validators[validatorId].reward = 0;
+    require(token.transfer(msg.sender, amount), "Insufficent rewards");
+  }
+
+  function delegationTransfer(uint256 amount, address delegator) public {
```
Comment (atvanguard) [inline contracts/staking/StakeManager.sol]: Where is this being used?

File: contracts/staking/StakeManager.sol (hunk 3 above)
Comment (0xAshish) [reply contracts/staking/StakeManager.sol]: hoping this never comes, should I add `pos 2/3+1 has colluded` to the require msg?

File: contracts/staking/StakeManager.sol [hunk 5]
```diff
@@ -28,17 +31,24 @@ contract StakeManager is Validator, IStakeManager, RootChainable, Lockable {
   uint256 public EPOCH_LENGTH = 256; // unit : block
   uint256 public UNSTAKE_DELAY = DYNASTY.mul(2); // unit: epoch
 
+  // TODO: add events and gov. based update function
+  uint256 public proposerToSignerRewards = 10; // will be used with fraud proof
```
[the remark this answers is not in the document]
Comment (0xAshish) [reply contracts/staking/StakeManager.sol]: Heimdall is picking this ratio from the contract and later on, it'll also be used in `challangeStateRootUpdate`.

File: contracts/staking/StakeManager.sol [hunk 6]
```diff
@@ -28,17 +31,24 @@ contract StakeManager is Validator, IStakeManager, RootChainable, Lockable {
...
+  uint256 public totalRewards;
+  uint256 public totalRewardsLiquidated;
+  bytes32 public accountStateRoot;
 
   enum Status { Inactive, Active, Locked }
 
   struct Validator {
     uint256 amount;
     uint256 reward;
+    uint256 totalReward;
```
[the remark this answers is not in the document]
Comment (0xAshish) [reply contracts/staking/StakeManager.sol]: done

File: contracts/staking/StakeManager.sol (hunk 4 above)
Comment (0xAshish) [reply contracts/staking/StakeManager.sol]: ah! forgot to use it in delegation manager.

Comment (0xAshish) [reply contracts/staking/StakeManager.sol]: Thanks

File: contracts/staking/StakeManager.sol [hunk 7]
```diff
@@ -165,6 +176,40 @@ contract StakeManager is Validator, IStakeManager, RootChainable, Lockable {
...
+    if (_contract == address(0x0)) {
+      validators[validatorId].reward = validators[validatorId].reward.add(_reward);
+    }
+    else {
+      // TODO: delegator bond/share rate if return needs to be updated periodically
+      // otherwise validator can delay and get all the delegators reward
+      ValidatorContract(_contract).updateRewards(_reward, currentEpoch, validators[validatorId].amount);
+    }
+    totalRewardsLiquidated += _reward;
+    require(totalRewardsLiquidated <= totalRewards, "Liquidating more rewards then checkpoints submitted");// pos 2/3+1 is colluded
```
Comment (atvanguard) [inline contracts/staking/StakeManager.sol]: Lol, it's still incorrect. Nevermind.

File: contracts/staking/StakeManager.sol (hunk 3 above)
Comment (atvanguard) [reply contracts/staking/StakeManager.sol]: Lol, it's still incorrect. Nevermind.

**Pull request merged**